### ISSUE TITLE
Drawer doesn't hide (fully open)

### ISSUE THREAD

**Issue by Ryszard-Trojnacki** [rmwc/rmwc#1211]
- What RMWC Version are you using [major.minor.patch]:
`14.0.3`

- Name your build system [Webpack, Rollup...]:
Not sure. It is React 18.2.0.

- Describe the bug with as much detail as possible:

I'm using **RMWC** version **14.0.3** (`React 18.2.0`) and I have a problem with `Drawer` component which is not closing/hidding.

The problem is on web mode using Google Chrome and also on Apache Cordova eigher on emulator and physical phone.

I have spend some time on this and it seems that the problem is with bad opening drawer rather than closing.
If drawer is badly open then it doesn't want to close.

I have checked the value of property `open` and it is correct (set to `false`) but drawer is visible. This not happens every time but only when opening is not complete.

Sometimes (to often) when I open drawer it ends with state opening.
Valid open of drawer ends with tag `aside` with classes: `mdc-drawer--open mdc-drawer mdc-drawer--modal`, but once in a few attemps it stops at state `opening` with classes set to: `mdc-drawer--open mdc-drawer--animate mdc-drawer--opening mdc-drawer mdc-drawer--modal` and stays with this. When it is in this state setting property `open` to `false` doesn't do anything.

Example application is at: https://rmwc.io/drawers you have to press `Toggle Modal` button few (about 1-10 times to repeat).

![image](https://github.com/rmwc/rmwc/assets/7201935/30685d90-acc2-4fa2-b9ef-fcc96853b5dc)

Tested on :
- Windows 11 with Google Chrome 119.0.6045.200, 
- Windows 11 with Microsoft Edge 119.0.2151.93, 
- Debian with Google Chrome  119.0.6045.159, 
- Debian with Firefox 115.5.0esr, 
- Android Emulator as as a Apache Cordova application, 
- Samsung Galaxy S10e phone as a Apache Cordova application. Apache Codrova Android plugin is `12.0.1`.

**Comment by Ryszard-Trojnacki** [comment]
I have dig into this issue more and problem is (I think, not 100% sure) with adding class `mdc-drawer--opening` it is added to fast (?). In the same time as `mdc-drawer--animate` and animation is not executed. Drawer is displayed fully open without animation. This causes that `transition end` [handler](https://github.com/material-components/material-components-web/blob/d76666ad4c53a619946d70b3991ca35ae5d87d04/packages/mdc-drawer/dismissible/foundation.ts#L144) is not called and everything later is not working. Component is in invalid state. [Close](https://github.com/material-components/material-components-web/blob/d76666ad4c53a619946d70b3991ca35ae5d87d04/packages/mdc-drawer/dismissible/foundation.ts#L98) method  is checking state at begining and it is invalid.

Drawer opening is executed from `@material` library from file [foundation.ts](https://github.com/material-components/material-components-web/blob/d76666ad4c53a619946d70b3991ca35ae5d87d04/packages/mdc-drawer/dismissible/foundation.ts#L77):
```javascript
  open() {
    if (this.isOpen() || this.isOpening() || this.isClosing()) {
      return;
    }

    this.adapter.addClass(cssClasses.OPEN);
    this.adapter.addClass(cssClasses.ANIMATE);

    // Wait a frame once display is no longer "none", to establish basis for
    // animation
    this.runNextAnimationFrame(() => {
      this.adapter.addClass(cssClasses.OPENING);
    });

    this.adapter.saveFocus();
  }
```

This above has only one connection to `@rmwc` library, that is file [drawer/src/lib/foundation.tsx](https://github.com/rmwc/rmwc/blob/b0d9fab9227be2e5e0b29f2353abe82bf760df3c/packages/drawer/src/lib/foundation.tsx#L1):
```javascript
        const f = new MDCConstructor({
          addClass: (className: string) => rootEl.addClass(className),
          removeClass: (className: string) => rootEl.removeClass(className),
          hasClass: (className: string) => rootEl.hasClass(className),
```

Code `this.adapter.addClass(cssClasses.OPENING);` will call `addClass` from `@rmwc` library.

I have made a not to good (in terms of resolving this in library) workaround but it works. I have changed method `addClass` and it helped:
```javascript
        addClass: (s) => {
          if(s==='mdc-drawer--opening') window.setTimeout(() => o.addClass(s), 30);
          else o.addClass(s);
```
I'm just delaying call for adding class of `mdc-drawer--opening`.

I hope this will help with resolving this issue.

**Comment by jensborch** [comment]
@Ryszard-Trojnacki Thank you for reporting. We will have a look at the issue and try to resolve it

**Comment by jensborch** [comment]
@Ryszard-Trojnacki  I am having a hard time reproducing this error. I have only managed to do it once. This makes it difficult to finde a solution. Also, this actually seems like a bug in Material and not RMWC, but we will of course be happy to add a workaround if needed.

I am wondering if having timout of 0 (similar to what is done her https://github.com/material-components/material-components-web/blob/d76666ad4c53a619946d70b3991ca35ae5d87d04/packages/mdc-drawer/dismissible/foundation.ts#L185)

´´´
          addClass: (className: string) => { 
            if(className==='mdc-drawer--opening') {
              window.setTimeout(() => rootEl.addClass(className), 0);
           }
```

or using requestAnimationFrame will fix the issue

´´´
          addClass: (className: string) => { 
            if(className==='mdc-drawer--opening') {
              window.requestAnimationFrame(() => {
                rootEl.addClass(className);
              });
            } 
```

Could you test this for me?

**Comment by Ryszard-Trojnacki** [comment]
@jensborch Before submitting this issue I have checked this component directly on material-web demo:
https://material-components.github.io/material-components-web-catalog/#/component/drawer/modal
I wasn't able to reproduce this bug. Checked few times. Checked once again before writing this comment.

On https://rmwc.io/drawers I have 100% success rate on different system and browers. Checked once again now.

I have attached video of reproducing this issue two times today.
https://youtu.be/8Xtl1EbXbcs
https://youtu.be/SwEEoXOBMlA

I will check your solution tomorrow.

I'm not sure if `window.requestAnimationFrame` will work, because:
```javascript
this.runNextAnimationFrame(() => {
      this.adapter.addClass(cssClasses.OPENING);
    });
```
this is already called in `nextAnimationFrame`.
